apartment/utils.py

```diff
@@ -17,6 +20,37 @@ def get_apartment_state_from_apartment_uuid(apartment_uuid):
     ).value
 
 
+def get_apartment_state_of_sale_from_event(event):
+    """
+    If there is a reservation marked as Sold in Sales UI, the apartment state of sale
+    should be also changed to SOLD
+    If the apartment has no reservations in the sales tool it should show as
+    vacant (vapaa) on the Drupal site
+    If it has any reservations - regardless the status of reservations -
+    the apartment should show as RESERVED (varattu) or RESERVED_HASO (käsittelyssä)
+    depending on the apartment type
+    """
+    if event.state == ApartmentReservationState.SOLD:
+        return ApartmentStateOfSale.SOLD
+    # Should only check for `FREE` state if
+    # the latest change is a reservation cancellation
+    if (
+        event.state == ApartmentReservationState.CANCELED
+        and ApartmentReservation.objects.active()
+        .filter(apartment_uuid=event.reservation.apartment_uuid)
+        .count()
+        == 0
+    ):
+        return ApartmentStateOfSale.FREE_FOR_RESERVATIONS
+    else:
+        apartment_type = get_apartment(
+            event.reservation.apartment_uuid, include_project_fields=True
+        ).project_ownership_type
+        if apartment_type.lower() == OwnershipType.HASO.value:
+            return ApartmentStateOfSale.RESERVED_HASO
+    return ApartmentStateOfSale.RESERVED
```

Comment:
It might be just me so feel free to ignore, but I find this if/else thing a bit confusing when there is also the bottom "default" return value. What about
```suggestion
    if (
        event.state == ApartmentReservationState.CANCELED
        and ApartmentReservation.objects.active()
        .filter(apartment_uuid=event.reservation.apartment_uuid)
        .count()
        == 0
    ):
        return ApartmentStateOfSale.FREE_FOR_RESERVATIONS

    apartment_type = get_apartment(
        event.reservation.apartment_uuid, include_project_fields=True
    ).project_ownership_type
    if apartment_type.lower() == OwnershipType.HASO.value:
        return ApartmentStateOfSale.RESERVED_HASO
    else:
        return ApartmentStateOfSale.RESERVED
```

Reply:
Also, I think there is one edge case where this returns a wrong state; if an apartment's first reservation is in state SOLD, and then you cancel some other reservation from the same apartment, then the apartment's state will be returned as RESERVED instead of SOLD. I **think** 🙂.

Ofc you wouldn't normally go and cancel reservations from an apartment that is already sold, but that is at least possible to do in the UI.

Reply:
> Also, I think there is one edge case where this returns a wrong state; if an apartment's first reservation is in state SOLD, and then you cancel some other reservation from the same apartment, then the apartment's state will be returned as RESERVED instead of SOLD. I **think** 🙂.
> 
> Ofc you wouldn't normally go and cancel reservations from an apartment that is already sold, but that is at least possible to do in the UI.

Nice catch! I fixed the logic to handle that case